infrasim/model/elements/backend.py

```diff
@@ -1,117 +1,140 @@
+        super(CBackendStorage, self).__init__()
+        self.__backend_storage_info = backend_storage_info
+        self.__controller_list = []
+        self.__pci_topology_manager = None
+
+        # Global controller index managed by CBackendStorage
+        self.__sata_controller_index = 0
+        self.__scsi_controller_index = 0
+        self.__nvme_controller_index = 0
+
+    def set_pci_topology_mgr(self, ptm):
+        self.__pci_topology_manager = ptm
+
+    def precheck(self):
+        for controller_obj in self.__controller_list:
+            controller_obj.precheck()
+
+    def __create_controller(self, controller_info):
+        controller_obj = None
+        model = controller_info.get("type", "ahci")
+        if model.startswith("megasas"):
+            controller_obj = MegaSASController(controller_info)
+        elif model.startswith("lsi"):
+            controller_obj = LSISASController(controller_info)
+        elif "nvme" in model:
+            controller_obj = NVMeController(controller_info)
+        elif "ahci" in model:
+            controller_obj = AHCIController(controller_info)
+        else:
+            raise ArgsNotCorrect("[BackendStorage] Unsupported controller type: {}".
+                                 format(model))
+
+        controller_obj.logger = self.logger
+        # set owner
+        controller_obj.owner = self
+        return controller_obj
+
+    def __get_ws_name(self):
+        parent = self.owner
+        while parent and not hasattr(parent, "get_workspace"):
+            parent = parent.owner
+
+        ws = None
+        if hasattr(parent, "get_workspace"):
+            ws = parent.get_workspace()
+
+        if ws is None or not os.path.exists(ws):
+            ws = ""
+
+        return ws
+
+    def init(self):
+        # parse disk array definition in advance.
+        dae_helper = DAEProcessHelper(self.__backend_storage_info,
+                                      self.__get_ws_name())
+
+        for controller in self.__backend_storage_info:
+            controller_obj = self.__create_controller(controller)
+            if self.__pci_topology_manager:
+                controller_obj.set_pci_topology_mgr(self.__pci_topology_manager)
+            self.__controller_list.append(controller_obj)
+
+        for controller_obj in self.__controller_list:
+            if isinstance(controller_obj, AHCIController):
+                controller_obj.controller_index = self.__sata_controller_index
+            elif isinstance(controller_obj, NVMeController):
+                controller_obj.controller_index = self.__nvme_controller_index
+            else:
+                controller_obj.controller_index = self.__scsi_controller_index
+
+            controller_obj.init()
+            # generate and save drives and traversal information.
+            if isinstance(controller_obj, CBaseStorageController):
+                dae_helper.create_dae_node(controller_obj.get_drvs_arg_option())
```

Comment:
Key point ⭐️ 1: get options of storage controllers and drives from yml definition, and inject to the DAE model. Then `dae_helper.create_dae_node` will dump this information to a file defined by `dae_file` in yml. As a result, qemu options take this dae file and parse everything inside qemu.

Reply:
Bingo 💯